app/lib/efile/nj/nj1040_calculator.rb

```diff
@@ -319,20 +338,24 @@ def calculate_line_16b
         calculate_tax_exempt_interest_income if calculate_tax_exempt_interest_income.positive?
       end
 
-      def calculate_line_20a
-        non_military_1099rs.sum(&:taxable_amount).round
-      end
-
       def calculate_line_20b
         (non_military_1099rs.sum(&:gross_distribution_amount) - non_military_1099rs.sum(&:taxable_amount)).round
       end
 
-      def calculate_line_27
-        line_or_zero(:NJ1040_LINE_15) + line_or_zero(:NJ1040_LINE_16A) + line_or_zero(:NJ1040_LINE_20A)
+      def calculate_line_28b
+        helper = Efile::Nj::NjRetirementIncomeHelper.new(@intake)
+        return 0 unless helper.eligible?
+        
+        total_income = calculate_line_27
+        [helper.calculate_maximum_exclusion(total_income) - calculate_line_28a, helper.total_eligible_nonretirement_income].min
+      end
+
+      def calculate_line_28c
+        calculate_line_28a + calculate_line_28b
       end
 
       def calculate_line_29
-        line_or_zero(:NJ1040_LINE_27)
+        line_or_zero(:NJ1040_LINE_27) - calculate_line_28c
```

Comment:
[dust] Why not use the line_or_zero instead of calculate_line_28c for consistency and to avoid calculating things more than once?